crates/marketplace/src/builder_state.rs

```diff
@@ -937,3 +948,227 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
+            .read_arc()
+            .await
+            .spawned_builder_states
+            .get(&builder_state_id)
+        {
+            panic!("global_state shouldn't have cooresponding builder_state_id without matching quorum proposal.");
+        }
+
+        // sub-test two
+        // call process_da_proposal with the same msg again
+        // we should skip the process and everything should be the same
+        let transactions_1 = transactions.clone();
+        let (_quorum_proposal_1, _quorum_proposal_msg_1, da_proposal_msg_1, builder_state_id_1) =
+            calc_proposal_msg(NUM_STORAGE_NODES, 0, None, transactions_1).await;
+        if let MessageType::DaProposalMessage(practice_da_msg_1) = da_proposal_msg_1.clone() {
+            builder_state
+                .process_da_proposal(practice_da_msg_1.clone())
+                .await;
+        } else {
+            tracing::error!("Not a da_proposal_message in correct format");
+        }
+        check_equal_da_proposal_hashmap(
+            builder_state
+                .da_proposal_payload_commit_to_da_proposal
+                .clone(),
+            correct_da_proposal_payload_commit_to_da_proposal.clone(),
+        );
+        // check global_state didn't change
+        if let Some(_x) = global_state
+            .read_arc()
+            .await
+            .spawned_builder_states
+            .get(&builder_state_id_1)
+        {
+            panic!("global_state shouldn't have cooresponding builder_state_id without matching quorum proposal.");
+        }
+
+        // sub-test three
+        // add the matching quorum proposal message with different tx
+        // and call process_da_proposal with this matching da proposal message and quorum proposal message
+        // we should spawn_clone here
+        // and check whether global_state has correct BuilderStateId
+        let transactions_2 = vec![TestTransaction::new(vec![1, 2, 3, 4]); 2];
+        let (_quorum_proposal_2, quorum_proposal_msg_2, da_proposal_msg_2, builder_state_id_2) =
+            calc_proposal_msg(NUM_STORAGE_NODES, 0, None, transactions_2).await;
+
+        // process quorum proposal first, so that later when process_da_proposal we can directly call `build_block` and skip storage
+        if let MessageType::QuorumProposalMessage(practice_qc_msg_2) = quorum_proposal_msg_2.clone()
+        {
+            builder_state
+                .process_quorum_proposal(practice_qc_msg_2.clone())
+                .await;
+        } else {
+            tracing::error!("Not a quorum_proposal_message in correct format");
+        }
+
+        // process da proposal message and do the check
+        if let MessageType::DaProposalMessage(practice_da_msg_2) = da_proposal_msg_2.clone() {
+            builder_state
+                .process_da_proposal(practice_da_msg_2.clone())
+                .await;
+        } else {
+            tracing::error!("Not a da_proposal_message in correct format");
```

Comment:
Nit: I think it's better to `panic` directly in tests to make it explicit that we hit something unexpected.

Reply:
Updated!